Components/LearningCore/Source/gov/sandia/cognition/learning/algorithm/clustering/OptimizedAffinityPropagation.java

```diff
@@ -0,0 +1,318 @@
+     */
+    @Override
+    protected void updateResponsibilities()
+    {
+
+      for(int i = 0 ; i < exampleCount; i++) {
+        double[] rowArray = createSumArray(availabilities[i], similarities[i]);
+        MaxResult maxResult = computeMax(rowArray);
+        for (int j = 0; j < exampleCount; j++) {
+          double maxToSubtract = j != maxResult.maxIndex ? maxResult.maxValue : maxResult.secondMaxValue;
+          responsibilities[i][j] = oneMinusDampingFactor * (similarities[i][j] - maxToSubtract) + dampingFactor * responsibilities[i][j];
+        }
+      }
+    }
+
+    private MaxResult computeMax(double[] rowArray)
+    {
+      MaxResult maxResult = new MaxResult();
+      for (int i = 0; i < rowArray.length; i++) {
+        if(rowArray[i] > maxResult.maxValue){
+          maxResult.secondMaxValue = maxResult.maxValue;
+          maxResult.maxValue = rowArray[i];
+          maxResult.maxIndex = i;
+        }else if(rowArray[i] > maxResult.secondMaxValue){
+          maxResult.secondMaxValue = rowArray[i];
+        }
+      }
+      return maxResult;
+    }
+
+
+    private double[] createSumArray(double[] availability, double[] similarity)
+    {
+      double[] resultArray = new double[availability.length];
+      for (int i = 0; i < availability.length; i++) {
+        resultArray[i] = availability[i] + similarity[i];
+      }
+      return resultArray;
+    }
+
+    /**
+     * Updates the availabilities matrix based on the current responsibility
+     * values.
+     */
+    @Override
+    protected void updateAvailabilities()
+    {
+      ColumnArray rp = new ColumnArray(exampleCount);
+      for (int j = 0; j < exampleCount; j++) {
+        removeNegativeValues(rp, responsibilities, j);
+        rp.array[j] = responsibilities[j][j];
+        rp.sum += responsibilities[j][j];
+        updateAvailabilitiesMatrix(j, rp, availabilities);
+      }
+    }
+
+    private void updateAvailabilitiesMatrix(int j, ColumnArray rp, double[][] availabilities)
+    {
+      double old;
+      for (int i = 0; i < availabilities[0].length; i++) {
+        old = availabilities[i][j];
+        availabilities[i][j] = rp.sum - rp.array[i];
+        if(i != j){
+          availabilities[i][j] = Math.min(availabilities[i][j], 0);
+        }
+        availabilities[i][j] = oneMinusDampingFactor * availabilities[i][j] + dampingFactor * old;
+      }
+    }
+
+    private void removeNegativeValues(ColumnArray columnArray, double[][] matrix, int j)
+    {
+      columnArray.sum = 0;
+      for (int i = 0; i < matrix.length; i++) {
+        columnArray.array[i] = Math.max(matrix[i][j], 0);
+        columnArray.sum += columnArray.array[i];
+      }
+    }
+
+    class MaxResult {
+      int maxIndex;
+      double maxValue;
+      double secondMaxValue;
+
+      MaxResult(){
+        maxValue = Double.NEGATIVE_INFINITY;
+        secondMaxValue = Double.NEGATIVE_INFINITY;
+        maxIndex = -1;
+      }
+    }
+
+    class ColumnArray

```

Comment:
Rather than needing a separate class for ColumnArray I think you can use a Vector (and then cache the call to sum() in the method that uses it above).

Reply:
I hope the modification i've done is what you expected